extensions/resteasy/deployment/src/main/java/io/quarkus/resteasy/deployment/ResteasyScanningProcessor.java

```diff
@@ -309,13 +316,18 @@ public void build(
 
             Set<String> resources = new HashSet<>();
             Set<DotName> pathInterfaces = new HashSet<>();
+            Set<String> withoutDefaultCtor = new HashSet<>();
             for (AnnotationInstance annotation : paths) {
                 if (annotation.target().kind() == AnnotationTarget.Kind.CLASS) {
                     ClassInfo clazz = annotation.target().asClass();
                     if (!Modifier.isInterface(clazz.flags())) {
                         String className = clazz.name().toString();
                         resources.add(className);
                         reflectiveClass.produce(new ReflectiveClassBuildItem(true, true, className));
+
+                        if (!clazz.hasNoArgsConstructor()) {
```

Comment:
I'm quite sure we can do this for every resource class. JAX-RS does support some kind of constructor injection, or? Actually, I found a bug in our ArC integration - the following constructor results in unsatisfied dependency:
```java
public TestResourceWithConstructorInjection(@Context Application application) {
   this.application = application;
}
```
Because we automatically consider this a CDI bean constructor.

Reply:
Hm, by the way if I annotate the class with `@ApplicationScoped` I get: 
```
javax.enterprise.inject.spi.DefinitionException: Normal scoped beans must declare a non-private constructor with no parameters
```
Because ArC is analyzing the original class (jandex view) and not the modified one.

Reply:
Finally, the generated no-arg constructor would not be usable - it does not initialize the final fields. So I would expect some ugly exception when it's used...

Reply:
I think I am missing some context here. I there a reason one would annotate a `@Path` resource with `@ApplicationScoped`?

The tests currently pass with the `final` fields, there is no problem at all. I haven't tried with `@ApplicationScoped` however.

Also, I didn't really understand the first comment, do you mean "I',m *not* quite sure"? 

So could this all just boil down to the following: Only generate a default constructor when there is no `@ApplicationScoped` annotation (and perhaps a few other constraints)?

Reply:
> I there a reason one would annotate a @Path resource with @ApplicationScoped?

It is allowed. In fact, in CDI world `@ApplicationScoped` is usually used instead of `@Singleton`.

> The tests currently pass...

Yep, I think it passes because the constructor is not used at all.

> Also, I didn't really understand the first comment...

Ah, sorry, yes it should be: "I'm NOT quite sure "

> So could this all just boil down to the following...

I'm not sure. We should think about it...

Reply:
My idea is to make the common case easy for developes - meaning that when they write a JAX-RS resource they can use constuctor injection (this is basically what Spring MVC users expect). 
So I would propose that we find the most restricted set of use cases where this is possible ( for example no `@ApplicationScoped` and I don't know what else :) ) and then later on if possible lift these restrictions. WDYT?

Reply:
Is it that bad to require a constructor be added?

Trying to see what the benefits are, and whether they're outweighed by adding "magic" all over the place

Reply:
@kenfinnigan this is the first thing that my local JUG leader (and Java Champion Paris Apostolopoulos) complained about for Quarkus. Spring folks are soooooooo used to not needing the extra boilerplate.

Reply:
Ok, but I don't necessarily like the idea of sprinkling magic all over the place just to get Spring folks to look at Quarkus

Reply:
It's only magic on the implementation side. For a user perspective having constructor injection work all over the place, makes perfect sense I think. For example, Arc supports constructor injection OOTB, so why not have this support everywhere?

Reply:
As we're trying to remove code as much as possible to reduce size, it's more a concern about this type of thing being added all over the place in Quarkus to support a specific use case, potentially at the detriment of others.

Now that may not be the case here. I'm just cautioning against adding "magic" everywhere without good cause

Reply:
:+1: for not adding stuff were we don't need it! 
I am no expert in what overhead some constructors would add, but I would _guess_ that it wouldn't be too much - especially if we only add it under very certain circumstances. In this case I think the added developer joy, would warrant sprinkling a tiny bit of magic.

Reply:
Well, we're adding default constructors to JPA entities in `Panache` and we've already noticed that they're also useful for JSON entities. All this could be done in a new `RESTEasy with Panache` extension, but TBH I don't see how we can justify requiring the user write this constructor when it will be exactly the same as the generated one and it's required and enough to fix these use-cases. That seems to me to be the very definition of "boilerplate".

Reply:
I don't mind adding these constructors but they need to be valid, i.e. if a class declares a final field the constructor must initialize it to a default value as defined by JLS. 

The other problem is with re-indexing the modified class. I don't know how exactly do our bytecode transformers work but I suppose we don't replace the original `ClassInfo` from the index of a pre-indexed dependency? (which is one of the protomoted ways to add a dependency to the application index)

Reply:
@mkouba Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail? Also, do we need to worry about indexing the class? Are we really gonna need to know about the synthetic no-arg constructor anywere?

Reply:
> Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail?

Have you tried to use the generated constructor? If you only want to fool validation logic of some framework the it's ok but for example in case of CDI we do use the no-args constructor to instantiate client proxies. 

> Also, do we need to worry about reindexing.

ArC validation is based on jandex. So if there is a bean class that has normal scope and does not declare a no-args constructor the validation should fail (no matter what the real class look like).

Reply:
> > Is it even worth the trouble to generate a proper constructor (instead of just default one) even if the tests never fail?
> 
> Have you tried to use the generated constructor? If you only want to fool validation logic of some framework the it's ok but for example in case of CDI we do use the no-args constructor to instantiate client proxies.
> 

In this case it seems we are simply fooling RESTEasy - see [this](https://issues.jboss.org/browse/RESTEASY-2183?focusedCommentId=13715781). 

The more I think about it the more I think we should only be doing this if the resource is *not* annotated with `@ApplicationScoped` in order to avoid any suprises.
There should probably also be a restriction that there should not be a superclass other than `Object`.  

> > Also, do we need to worry about reindexing.
> 
> ArC validation is based on jandex. So if there is a bean class that has normal scope and does not declare a no-args constructor the validation should fail (no matter what the real class look like).

So if we only do this constructor generation for `Singleton` scope we can get around this restriction as well?

Reply:
I pushed a change that now only adds the default constuctor if there is no other annotation other than `@Path` and the resource doesn't extend from any class - This is should be the strictect possible (but most common I would say) use case.
@mkouba Do you think it is safe now?

Reply:
> So if we only do this constructor generation for Singleton scope we can get around this restriction as well?

Yes, `@Singleton` does not require a client proxy.

> Do you think it is safe now?

Well, the constructor would still not be valid if a class contains some final fields but it should be safe enough to fool RESTEasy ;-).

Reply:
My thoughts exactly :)

Reply:
@mkouba WDYT, should we merge?

Reply:
I'm +0. It looks like a lot of code to support only a very specific simple use case. @gsmet @kenfinnigan WDYT?

Reply:
One last thing if I may say so: The use case maybe simple and specific, but it goes a long way to create a consistent experience IMHO. Users should not have to add boilerplate (if that can be avoided) because of limitation of framework - they honestly don't care. Think of it this way: Why should a JAX-RS resource be any different from a user perspective than all the `Singleton` beans he/she adds to their app that work perfectly with constructor injection without requiring a no-arg constructor?